src/radar_control.py

```diff
@@ -35,6 +35,49 @@
 TIME_PROFILE = False
 
 
+@dataclass
+class CFSParameters:
+    """
+    Parameters used to track clear frequency search
+    """
+
+    cfs_freq: list = field(default_factory=list)
+    cfs_mags: list = field(default_factory=list)
+    cfs_range: list = field(default_factory=dict)
+    cfs_masks: list = field(default_factory=dict)
+    last_cfs_set_time: int = 0
+    last_cfs_power: list = field(default_factory=dict)
+    set_new_freq: bool = True
+
+    def check_update_freq(self, cfs_spectrum, cfs_slices, threshold):
+        """
+        Checks if any scanned frequencies have power levels that
+        exceed the current power of each cfs slice based on a threshold
+
+        :params cfs_spectrum:   Results of the CFS analysis
+        :type   cfs_spectrum:   dataclass
+        :params cfs_slices:     Slice ids of each cfs slice to be checked
+        :type   cfs_slices:     list
+        :params threshold:      Power threshold (dB) used in check
+        :type   threshold:      float
+        """
+        cfs_data = [dset.cfs_data for dset in cfs_spectrum.output_datasets]
+        for i, slice_id in enumerate(cfs_slices):
+            if any(
+                np.asarray(cfs_data[i][self.cfs_masks[slice_id]])
+                <= self.last_cfs_power[slice_id][1] - threshold
+            ):
+                self.set_new_freq = True
+            shifted_frequency = self.last_cfs_power[slice_id][0] - int(
+                sum(self.cfs_range[slice_id]) / 2
+            )
+            idx = (
+                np.abs(np.asarray(self.cfs_freq) - shifted_frequency * 1000)
+            ).argmin()
```

Comment:
I'm uncertain what the frequency shifting is for. Is this just computing the index into the frequency array of the currently-used frequency?

Reply:
yes